adding custom world presets (#79)
QuiltMC/developer-wiki: wiki/misc/worldtypes.md

---
title: Adding a Custom World Preset
---

# Adding custom world presets
If your mod completely overhauls Minecraft's worldgen, you'll want to create a custom worldtype so you give people the option to use your worldgen or MC's vanilla worldgen.

> Now, what is a custom worldtype you may ask? A custom worldtype is what will show up here: ![createANewWorld.png](createANewWorld.png)

Reviewer: 2 things!
- it's super nice to provide an image here, but it's formatted differently than the other ones. check `get-started.md` in the introduction category to see how to properly add an image
- this explains how to select a world type, not what a world type actually is!

Author: "Worldtypes are worldgen options for Minecraft worlds. What they do is change MC's worldgen based on what's defined in a file found in the `minecraft/worldgen/world_preset` directory."

Reviewer 2: how about something like "World types let you change the markup of your Minecraft worlds by configuring how world generation behaves"?

Author: > "the markup of your Minecraft worlds"
not entirely sure what "markup" would mean in this context

Reviewer 2: *makeup, sorry, typo'd

Author: > "World types are configurable world generation options for Minecraft worlds. They let you change the markup of your Minecraft worlds by configuring how world generation behaves."

how's this

To add your custom world preset(s), you have to navigate to your mod/datapack's `resources/data` folder.

Make a file titled `example_mod_preset.json` in `../example_mod/worldgen/world_preset` and populate it with the following:
```json
{
  "dimensions": {
    "minecraft:overworld": {
      "type": "minecraft:overworld",
      "generator": {
        "type": "minecraft:noise",
        "biome_source": {
          "type": "minecraft:multi_noise",
          "preset": "minecraft:overworld"
        },
        "settings": "example_mod:example_mod_noise"
      }
    }
  }
}
```
- `"dimensions`: The dimension(s) in your custom world preset. It is recommended to have at least `"minecraft:overworld"`, `"minecraft:nether"`, and `"minecraft:the_end"`.
- `"minecraft:overworld"`: This is required, otherwise your world preset will not work.

> - `"type"`: The namespace ID for the dimension you're using.

Author: ![image](https://github.com/QuiltMC/developer-wiki/assets/17211100/ee37c2de-96a8-42c5-b143-57e40e3bc92f)

> - `"type"`: The namespace ID for the dimension you're using.

Reviewer 3: I know which type since i looked at the minecraft wiki. I would not expect first time people looking at this to know which one it is.

- `"generator"`: Various settings to tell Minecraft how to generate the terrain found in this dimension.
- `"settings"`: What perlin noise to use for worldgen in this dimension.

> 

Reviewer 3: I would need to double check, but I would see if the dimension specification matches the normal dimension spec, and if so to consider making two articles for this

Make a file titled `normal.json` in `../minecraft/tags/worldgen/world_preset` and populate it with the following:
This will tell Minecraft to show your world preset when you go to create a new world.
```json
{
  "values": [

>     "wonders:wonders"

Reviewer: what is this key? we never mention "wonders" before this

Author: > what is this key? we never mention "wonders" before this

oh yea, that was because i copied the actual JSON from a [worldgen mod](https://github.com/DebuggyTeam/wonders/blob/1.20/src/main/resources/data/minecraft/tags/worldgen/world_preset/normal.json) i made and then just didn't continue development on

>     "example_mod:example_world_type"

Reviewer 3: ```suggestion
    "example_mod:example_mod_preset"
```

---

> title: Adding a Custom World Preset

Reviewer: after rereading a few times, the tone of this article is a bit off imo. it offers extremely short explanations of each thing (often without actually explaining anything. "Configurable noise generator." tells most people absolutely nothing), tosses pretty much all its information into one block in the middle instead of spacing it out across multiple code snippets like other wiki articles, and via those two vectors just has a generally terse tone, like it's pissed off at the user for not knowing about this and having to read the article

sorry to be mean, but I feel like this needs to be rewritten instead of getting bandaid fixes for review comments

---
title: Adding a Custom World Preset
---

# Adding custom world presets

> If your mod completely overhauls Minecraft's worldgen, you'll want to create a custom worldtype so you give people the option to use your worldgen or MC's vanilla worldgen.

Reviewer 3: ```suggestion
If your mod completely overhauls Minecraft's worldgen, you'll likely want to create a custom worldtype so you give people the option to use your worldgen or MC's vanilla worldgen.
```

> If your mod completely overhauls Minecraft's worldgen, you'll want to create a custom worldtype so you give people the option to use your worldgen or MC's vanilla worldgen.

Reviewer: ```suggestion
If your mod completely overhauls Minecraft's worldgen, you'll want to create a custom worldtype to offer users the option to use your worldgen or MC's vanilla worldgen.
```

Now, what is a custom worldtype you may ask? World types define world generation options for Minecraft worlds. They let you change the make-up of your Minecraft worlds by configuring how world generation behaves.

</br>
</br>

> Here is where you specify which dimensions your world preset will apply to. It does require a `minecraft:overworld` dimension or your world preset will not work.

Reviewer: this makes it sound like it specifically requires the `overworld` dimension, is that right?

> Valid dimensions are `minecraft:overworld`, `minecraft:nether` and `minecraft:the_end`.

Reviewer: as well as dimensions added by mods, I assume?

```json
"dimensions": {
  "minecraft:overworld": {

>     "type": "minecraft:overworld"

Reviewer 3: The type here isnt explained. Its the actual world type. In addition, the name of the object is the dimension name, and besides requiring `minecraft:overworld`, nothing else is. Make this distinction clearer.

    }
  }
```

</br>
</br>

`generator` is where you specify how the worldgen in your world preset behaves. The `type` is a "generator ID" and the following are valid generator IDs:
```
noise
flat
debug

> ```

Reviewer: explain these!

> `biome_source` is where you specify how biomes generate. The `type` parameter here is what kind of biome generation will be used.

Reviewer 3: This makes me think that i can add this to a flat or debug world, which is wrong. Make sure you a clear distinction of what fields are supported by what generator types.

Valid generation types:
```
multi_noise
fixed
checkerboard, if you want to play checkers in MC
the_end which generates similarly to MC's end dimension

> ```

Reviewer: explain all of these, and make sure that there's a clear separation between the id and the explanation (i'd use `"checkerboard",`)

If the biome source `type` happens to be `minecraft:multi_noise`, Minecraft will use 3D biomes in the overworld and the nether.
The `preset` can be `minecraft:overworld` or `minecraft:nether`. More info about this can be found [here](https://minecraft.wiki/w/Custom_dimension#Multi-noise_biome_source_parameter_list).

```json
"dimensions": {
  "minecraft:overworld": {
    "type": "minecraft:overworld",
    "generator": {
      "type": "minecraft:noise",
      "biome_source": {
        "type": "minecraft:multi_noise",
        "preset": "minecraft:overworld"
      }
    }
  }
}
```

</br>
</br>

If you specified `minecraft:flat` in the generator type, `settings` is where you'll configure your superflat generation settings. More info about world presets using `settings` can be found [here](https://minecraft.wiki/w/Custom_world_preset#World_presets_with_settings).
```json
"dimensions": {
  "minecraft:overworld": {
    "type": "minecraft:overworld",
    "generator": {

>       "type": "minecraft:noise",

Reviewer 3: should be `minecraft:flat`

      "biome_source": {
        "type": "minecraft:multi_noise",
        "preset": "minecraft:overworld"

>       },

Reviewer 3: not valid for a flat generator

>       "settings": 

Reviewer 3: missing values

>       "settings": 

Reviewer: empty?

    }
  }
}
```

</br>
</br>

```json
{
  "values": [
    "example_mod:example_mod_preset"
  ]
}
```

> In order for your world preset to show in the "create a new world" screen, you will need to make a file called `normal.json` in `resources/data/minecraft/tags/worldgen/world_preset` and populate it with the following.

Reviewer 3: This should have stayed above. I selected only the line i wanted changed

Reviewer 3: looks like github just showed a bunch of lines. sorry for the confusion

> Alternatively, if you want your world preset to only show when selecting world presets while holding down ALT, make a file called `extended.json` in the same directory as your `normal.json` file.

Reviewer 3: Move this below the code block.

> Alternatively, if you want your world preset to only show when selecting world presets while holding down ALT, make a file called `extended.json` in the same directory as your `normal.json` file.

Reviewer: ```suggestion
Alternatively, if you want your world preset to only show as an option while the user is holding down ALT, make a file called `extended.json` in the same directory as your `normal.json` file.
```
